python/lsst/obs/fiberspectrograph/spectrum.py

```diff
@@ -0,0 +1,198 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("FiberSpectrum",)
+
+import numpy as np
+import astropy.io.fits
+import fitsio
+import astropy.units as u
+from ._instrument import FiberSpectrograph
+import lsst.afw.image as afwImage
+
+
+class VisitInfo:
+    def __init__(self, md):
+        self.exposureTime = md["EXPTIME"]
+
+    def getExposureTime(self):
+        return self.exposureTime
+
+
+class Info:
+    def __init__(self, md):
+        self.visitInfo = VisitInfo(md)
+
+    def getVisitInfo(self):
+        return self.visitInfo
+
+
+class FiberSpectrum:
+    def __init__(self, wavelength, flux, md=None):
+        self.wavelength = wavelength
+        self.flux = flux
+        self.md = md
+
+        self.info = Info(md)
+        self.detector = FiberSpectrograph().getCamera()[0]
+
+        self.__Mask = afwImage.MaskX(1, 1)
+        self.getPlaneBitMask = self.__Mask.getPlaneBitMask  # ughh, awful Mask API
+        self.mask = np.zeros(flux.shape, dtype=self.__Mask.array.dtype)
+        self.variance = np.zeros_like(flux)
+
+    def getDetector(self):
+        return self.detector
+
+    def getInfo(self):
+        return self.info
+
+    def getMetadata(self):
+        return self.md
+
+    def getFilter(self):
+        return FiberSpectrograph.filterDefinitions[0].makeFilterLabel()
+
+    def getBBox(self):
+        return self.detector.getBBox()
+
+    @staticmethod
+    def readFits(path):
+        """Read a Spectrum from disk"
+
+        Parameters
+        ----------
+        path : `str`
+            The file to read
+
+        Returns
+        -------
+        spectrum : `~lsst.obs.fiberspectrograph.FiberSpectrum`
+            In-memory spectrum.
+        """
+        md = dict(fitsio.read_header(path))
+        flux = fitsio.read(path)
+        wavelength = fitsio.read(path, ext=md["PS1_0"], columns=md["PS1_1"]).flatten()
+
+        wavelength = u.Quantity(wavelength, u.Unit(md["CUNIT1"]), copy=False)
+
+        return FiberSpectrum(wavelength, flux, md)
+
+    def writeFits(self, path):
+        """Write a Spectrum to disk
+
+        Parameters
+        ----------
+        path : `str`
+            The file to write
+        """
+        hdl = DataManager(self).make_hdulist()
+
+        hdl.writeto(path)
+
+
+class DataManager:
```

Comment:
This looks like it's almost the same code from https://github.com/lsst-ts/ts_fiberspectrograph/blob/develop/python/lsst/ts/fiberspectrograph/data_manager.py

Are you required to keep this in sync with upstream if, say, FORMAT_VERSION changes?

Consider working with upstream to allow the upstream class to work here as well. It looks like things could be unified with some minor effort but I don't know if the output format here is required to match that written by the telescope and readable by the reader code (which is not looking at the FORMAT_V header).

Reply:
After discussing with @czwa, ts_fiberspectrograph not being in the stack makes this difficult. We suggest to leave this to another ticket.

Reply:
I wasn't saying call the class directly in ts_fiberspectrograph. I was suggesting that we work towards this package having a direct copy of the file so that it is clear where it came from. At minimum please move the data manager class into a separate file and document where it came from originally.

This package is a bit different to ts_fiberspectrograph in that the telescope package always writes the newest version but this obs package has to be able to read all the historical versions.